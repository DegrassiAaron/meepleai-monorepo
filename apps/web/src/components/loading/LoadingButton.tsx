/**
 * LoadingButton Component
 *
 * Button component with integrated loading state and spinner.
 * Automatically disables and shows loading indicator when isLoading is true.
 *
 * @example
 * ```tsx
 * const [isSubmitting, setIsSubmitting] = useState(false);
 *
 * const handleSubmit = async () => {
 *   setIsSubmitting(true);
 *   await submitForm();
 *   setIsSubmitting(false);
 * };
 *
 * <LoadingButton
 *   isLoading={isSubmitting}
 *   loadingText="Submitting..."
 *   onClick={handleSubmit}
 *   className="btn-primary"
 * >
 *   Submit Form
 * </LoadingButton>
 * ```
 */

import { ButtonHTMLAttributes, ReactNode } from 'react';
import { Spinner } from './Spinner';

export interface LoadingButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  /**
   * Whether the button is in loading state
   * @default false
   */
  isLoading?: boolean;

  /**
   * Text to display when loading (replaces children)
   */
  loadingText?: string;

  /**
   * Size of the loading spinner
   * @default 'md'
   */
  spinnerSize?: 'sm' | 'md' | 'lg';

  /**
   * Button content
   */
  children: ReactNode;
}

/**
 * LoadingButton component with integrated loading state
 */
export function LoadingButton({
  isLoading = false,
  loadingText,
  spinnerSize = 'md',
  children,
  className = '',
  disabled,
  ...buttonProps
}: LoadingButtonProps) {
  // Determine button disabled state
  const isDisabled = disabled || isLoading;

  return (
    <button
      type="button"
      {...buttonProps}
      disabled={isDisabled}
      aria-busy={isLoading}
      aria-live="polite"
      className={`
        inline-flex items-center justify-center gap-2
        transition-opacity duration-200
        ${isLoading ? 'opacity-70 cursor-not-allowed' : ''}
        ${className}
      `}
    >
      {isLoading && <Spinner size={spinnerSize} />}
      {isLoading ? loadingText || null : children}
    </button>
  );
}
