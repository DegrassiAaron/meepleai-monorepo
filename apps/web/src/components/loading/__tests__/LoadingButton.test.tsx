/**
 * Tests for LoadingButton component
 * Button with integrated loading state and spinner
 */

import { render, screen, fireEvent } from '@testing-library/react';
import { LoadingButton } from '../LoadingButton';

describe('LoadingButton', () => {
  describe('Basic rendering', () => {
    it('should render children when not loading', () => {
      render(<LoadingButton>Click me</LoadingButton>);
      expect(screen.getByRole('button')).toHaveTextContent('Click me');
    });

    it('should render as a button element', () => {
      render(<LoadingButton>Submit</LoadingButton>);
      const button = screen.getByRole('button');
      expect(button.tagName).toBe('BUTTON');
    });

    it('should apply default button type', () => {
      render(<LoadingButton>Submit</LoadingButton>);
      const button = screen.getByRole('button');
      expect(button).toHaveAttribute('type', 'button');
    });
  });

  describe('Loading state', () => {
    it('should show spinner when isLoading is true', () => {
      const { container } = render(<LoadingButton isLoading={true}>Submit</LoadingButton>);
      const spinner = container.querySelector('svg');
      expect(spinner).toBeInTheDocument();
    });

    it('should not show spinner when isLoading is false', () => {
      const { container } = render(<LoadingButton isLoading={false}>Submit</LoadingButton>);
      const spinner = container.querySelector('svg');
      expect(spinner).not.toBeInTheDocument();
    });

    it('should hide children when loading and no loadingText provided', () => {
      render(<LoadingButton isLoading={true}>Original Text</LoadingButton>);
      expect(screen.queryByText('Original Text')).not.toBeInTheDocument();
    });

    it('should show loadingText when provided and loading', () => {
      render(
        <LoadingButton isLoading={true} loadingText="Processing...">
          Submit
        </LoadingButton>
      );
      expect(screen.getByText('Processing...')).toBeInTheDocument();
      expect(screen.queryByText('Submit')).not.toBeInTheDocument();
    });

    it('should show children when not loading even with loadingText prop', () => {
      render(
        <LoadingButton isLoading={false} loadingText="Loading...">
          Submit
        </LoadingButton>
      );
      expect(screen.getByText('Submit')).toBeInTheDocument();
      expect(screen.queryByText('Loading...')).not.toBeInTheDocument();
    });
  });

  describe('Button disabled state', () => {
    it('should disable button when loading', () => {
      render(<LoadingButton isLoading={true}>Submit</LoadingButton>);
      expect(screen.getByRole('button')).toBeDisabled();
    });

    it('should not disable button when not loading', () => {
      render(<LoadingButton isLoading={false}>Submit</LoadingButton>);
      expect(screen.getByRole('button')).not.toBeDisabled();
    });

    it('should respect disabled prop when not loading', () => {
      render(<LoadingButton disabled={true}>Submit</LoadingButton>);
      expect(screen.getByRole('button')).toBeDisabled();
    });

    it('should be disabled when both disabled and isLoading are true', () => {
      render(
        <LoadingButton disabled={true} isLoading={true}>
          Submit
        </LoadingButton>
      );
      expect(screen.getByRole('button')).toBeDisabled();
    });
  });

  describe('Opacity styling', () => {
    it('should reduce opacity when loading', () => {
      render(<LoadingButton isLoading={true}>Submit</LoadingButton>);
      const button = screen.getByRole('button');
      expect(button).toHaveClass('opacity-70');
    });

    it('should have full opacity when not loading', () => {
      render(<LoadingButton isLoading={false}>Submit</LoadingButton>);
      const button = screen.getByRole('button');
      expect(button).not.toHaveClass('opacity-70');
    });
  });

  describe('Accessibility', () => {
    it('should have aria-busy attribute when loading', () => {
      render(<LoadingButton isLoading={true}>Submit</LoadingButton>);
      expect(screen.getByRole('button')).toHaveAttribute('aria-busy', 'true');
    });

    it('should have aria-busy=false when not loading', () => {
      render(<LoadingButton isLoading={false}>Submit</LoadingButton>);
      expect(screen.getByRole('button')).toHaveAttribute('aria-busy', 'false');
    });

    it('should have aria-live attribute', () => {
      render(<LoadingButton isLoading={true}>Submit</LoadingButton>);
      expect(screen.getByRole('button')).toHaveAttribute('aria-live', 'polite');
    });
  });

  describe('Event handlers', () => {
    it('should call onClick when not loading', () => {
      const handleClick = jest.fn();
      render(<LoadingButton onClick={handleClick}>Submit</LoadingButton>);
      fireEvent.click(screen.getByRole('button'));
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('should not call onClick when loading (button is disabled)', () => {
      const handleClick = jest.fn();
      render(
        <LoadingButton onClick={handleClick} isLoading={true}>
          Submit
        </LoadingButton>
      );
      fireEvent.click(screen.getByRole('button'));
      expect(handleClick).not.toHaveBeenCalled();
    });
  });

  describe('Spinner size', () => {
    it('should render spinner with custom size', () => {
      const { container } = render(
        <LoadingButton isLoading={true} spinnerSize="lg">
          Submit
        </LoadingButton>
      );
      const spinner = container.querySelector('svg');
      expect(spinner).toHaveAttribute('width', '32'); // lg = 32px
    });

    it('should default to medium spinner size', () => {
      const { container } = render(<LoadingButton isLoading={true}>Submit</LoadingButton>);
      const spinner = container.querySelector('svg');
      expect(spinner).toHaveAttribute('width', '24'); // md = 24px
    });
  });

  describe('Props forwarding', () => {
    it('should forward className to button', () => {
      render(<LoadingButton className="custom-class">Submit</LoadingButton>);
      expect(screen.getByRole('button')).toHaveClass('custom-class');
    });

    it('should forward type prop', () => {
      render(<LoadingButton type="submit">Submit</LoadingButton>);
      expect(screen.getByRole('button')).toHaveAttribute('type', 'submit');
    });

    it('should forward other HTML button attributes', () => {
      render(
        <LoadingButton data-testid="my-button" aria-label="Submit form">
          Submit
        </LoadingButton>
      );
      const button = screen.getByRole('button');
      expect(button).toHaveAttribute('data-testid', 'my-button');
      expect(button).toHaveAttribute('aria-label', 'Submit form');
    });
  });

  describe('Snapshot tests', () => {
    it('should match snapshot in default state', () => {
      const { container } = render(<LoadingButton>Submit</LoadingButton>);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot in loading state', () => {
      const { container } = render(
        <LoadingButton isLoading={true} loadingText="Processing...">
          Submit
        </LoadingButton>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
