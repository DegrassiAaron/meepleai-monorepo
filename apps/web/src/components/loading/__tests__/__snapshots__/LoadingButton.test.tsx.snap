// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`LoadingButton Snapshot tests should match snapshot in default state 1`] = `
<button
  aria-busy="false"
  aria-live="polite"
  class="
        inline-flex items-center justify-center gap-2
        transition-opacity duration-200
        
        
      "
  type="button"
>
  Submit
</button>
`;

exports[`LoadingButton Snapshot tests should match snapshot in loading state 1`] = `
<button
  aria-busy="true"
  aria-live="polite"
  class="
        inline-flex items-center justify-center gap-2
        transition-opacity duration-200
        opacity-70 cursor-not-allowed
        
      "
  disabled=""
  type="button"
>
  <div
    class="inline-block animate-spin "
  >
    <svg
      aria-hidden="true"
      fill="none"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        class="opacity-25"
        cx="12"
        cy="12"
        r="10"
        stroke="currentColor"
        stroke-width="4"
      />
      <path
        class="opacity-75"
        d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
        fill="currentColor"
      />
    </svg>
  </div>
  Processing...
</button>
`;
